File: online_armory/frontend/src/app/gun/gun.ts
import { Caliber } from "../caliber/caliber";
import { Company } from "../company/company";
import { GunType } from "../gyn-type/gun-type";
import { InternalMechanic } from "../internal-mechanic/internal-mechanic";
import { RearmingMode } from "../rearming-mode/rearming-mode";

export class Gun {
    id: number;
    company: Company;
    gunType: GunType;
    caliber: Caliber;
    rearmingMode: RearmingMode;
    internalMechanic: InternalMechanic;
    name: string;
    year: number;
    description: string;
    img: string;
    forbidden: boolean;
    free: boolean;
    barelSize: number;
    totalSize: number;

    constructor(
        company: Company = null,
        gunType: GunType = null,
        caliber: Caliber = null,
        rearmingMode: RearmingMode = null,
        internalMechanic: InternalMechanic = null,
        name: string = "",
        year: number = 0,
        description: string = "",
        img: string = "",
        forbidden: boolean = false,
        free: boolean = false,
        barelSize: number = 0,
        totalSize: number = 0,
    ){
        this.company = company;
        this.gunType = gunType;
        this.caliber = caliber;
        this.rearmingMode = rearmingMode;
        this.internalMechanic = internalMechanic;
        this.name = name;
        this.year = year;
        this.description = description;
        this.img = img;
        this.forbidden = forbidden;
        this.free = free;
        this.barelSize = barelSize;
        this.totalSize = totalSize;
    }
}

File: online_armory/backend/src/main/java/com/project/models/Gun.java
/*-----Gun-----
 * C'est le modèle d'une arme
 * Pour un modèle d'arme on a une marque, un type, un mode d'action, un ou plusieurs calibres
 */
package com.project.models;

import jakarta.persistence.*;
import lombok.NoArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Entity
@Getter
@Setter
@ToString
@NoArgsConstructor
public class Gun {

    @Id
    @GeneratedValue(strategy= GenerationType.AUTO)
    private Long id;
    
    //Un gun ne peut avoir qu'un seul fabricant, mais un fabriquant peut avoir plusieurs modèles
    @ManyToOne
    private Company company;//Fabricant

    //Un gun ne peut avoir qu'un seul type, mais un type peut être décliné en plusieurs modèles
    @ManyToOne
    private GunType gunType;//Revolver, Pistolet, Carabine ou Fusil

    //Un gun ne peut avoir qu'un seul mode de réarmement, mais un mode de réarmement peut être décliné en plusieurs modèles
    @ManyToOne
    private RearmingMode rearmingMode;//Mode de réarmement de l'arme

    @ManyToOne
    private InternalMechanic internalMechanic;

    //Un gun ne peut avoir qu'un seul calibre, mais un calibre peut être utilisé par plusieurs modèles
    @ManyToOne
    private Caliber caliber;//Calibre
    private String name;

    private String year;
    private String description;
    private String img;
    private boolean forbidden;//Est-ce une arme prohibée en Belgique ?
    private boolean free;//Est-ce une arme en vente libre en Belgique ?
    private int barelSize;//Taille du canon
    private int totalSize;//Taille totale

    public Gun(Company company, GunType gunType, RearmingMode rearmingMode, Caliber caliber, String name, String year, String description, String img, boolean forbidden, boolean free, int barelSize, int totalSize){
        this.company = company;
        this.gunType = gunType;
        this.rearmingMode = rearmingMode;
        this.caliber = caliber;
        this.name = name;
        this.year = year;
        this.description = description;
        this.img = img;
        this.forbidden = forbidden;
        this.free = free;
        this.barelSize = barelSize;
        this.totalSize = totalSize;
    }

    
    
}
